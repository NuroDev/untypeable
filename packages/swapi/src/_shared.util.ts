import { beforeAll } from "vitest";
import { createTypeLevelClient } from "untypeable";
import { fetch } from "undici";
import { join } from "node:path";

import type { SwapiRouter } from ".";

export function useTestClient() {
  const client = createTypeLevelClient<SwapiRouter>(
    async (path, input = {}) => {
      const url = new URL(join("/api", path), "https://swapi.dev/");
      Object.entries(input).forEach(([key, value]) =>
        url.searchParams.append(key, value as string)
      );

      const response = await fetch(url.href, {
        method: "GET",
        headers: {
          "Content-Type": "application/json",
        },
      });

      return await response.json();
    }
  );

  beforeAll(() => {
    if (!client) throw "Failed to initialise untypeable client instance.";
  });

  return client;
}
