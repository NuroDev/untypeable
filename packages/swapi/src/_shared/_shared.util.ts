import { beforeAll } from "vitest";
import { createTypeLevelClient } from "untypeable";
import { fetch } from "undici";
import { join } from "node:path";

import type { SwapiRouter } from "..";

export function useTestClient() {
  const client = createTypeLevelClient<SwapiRouter>(
    async (path, input = {}) => {
      const pathWithParams = path.replace(
        /:([a-zA-Z0-9_]+)/g,
        (_, key) => input[key]
      );

      const url = new URL(join("/api", pathWithParams), "https://swapi.dev/");

      const response = await fetch(url.href, {
        method: "GET",
        headers: {
          "Content-Type": "application/json",
        },
      });

      return await response.json();
    }
  );

  beforeAll(() => {
    if (!client) throw "Failed to initialise untypeable client instance.";
  });

  return client;
}
