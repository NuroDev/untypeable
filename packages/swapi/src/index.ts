import { initUntypeable } from "untypeable";

import type { Film, FilmParams, Films } from "./films.types";
import type { People, Person, PersonParams } from "./people.types";
import type { Root } from "./root.types";

const u = initUntypeable();

const router = u.router({
  "/": u.output<Root>(),
  /** Get all the film resources */
  "/films": u.output<Films>(),
  /** Get a specific film resource */
  "/films/:id": u.input<FilmParams>().output<Film>(),
  /** Get all the people resources */
  "/people": u.output<People>(),
  /** Get a specific people resource */
  "/people/:id": u.input<PersonParams>().output<Person>(),
});

export type SwapiRouter = typeof router;
