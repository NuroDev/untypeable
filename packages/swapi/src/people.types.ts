import type { z } from "zod";

import type {
  PeopleSchema,
  PersonParamsSchema,
  PersonSchema,
} from "./people.validators";

/**
 * @docs https://swapi.dev/documentation#people
 */
export type Person = z.infer<typeof PersonSchema>;

/**
 * @docs https://swapi.dev/documentation#people
 */
export type People = z.infer<typeof PeopleSchema>;

export type PersonParams = z.infer<typeof PersonParamsSchema>;
