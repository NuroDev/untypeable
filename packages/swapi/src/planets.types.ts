import type { z } from "zod";

import type {
  PlanetParamsSchema,
  PlanetSchema,
  PlanetsSchema,
} from "./planets.validators";

/**
 * @docs https://swapi.dev/documentation#planets
 */
export type Planet = z.infer<typeof PlanetSchema>;

/**
 * @docs https://swapi.dev/documentation#planets
 */
export type Planets = z.infer<typeof PlanetsSchema>;

export type PlanetParams = z.infer<typeof PlanetParamsSchema>;
