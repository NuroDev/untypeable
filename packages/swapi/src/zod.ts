export * from "./films.validators";
export * from "./people.validators";
export * from "./planets.validators";
export * from "./root.validators";
export * from "./species.validators";
export * from "./starships.validators";
export * from "./vehicles.validators";
