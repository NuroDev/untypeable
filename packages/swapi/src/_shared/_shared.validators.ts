import { z } from "zod";

export function defineSWAPIPaginatedSchema<TSchema extends z.ZodTypeAny>(
  dataSchema: TSchema
) {
  return z.object({
    count: z.number(),
    next: z.string().url().nullable(),
    previous: z.string().url().nullable(),
    results: z.array(dataSchema),
  });
}
