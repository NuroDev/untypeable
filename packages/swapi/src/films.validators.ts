import { z } from "zod";
import { defineSWAPIPaginatedSchema } from "./_shared/_shared.validators";

/**
 * @docs https://swapi.dev/documentation#films
 */
export const FilmSchema = z.object({
  /**
   * An array of people resource URLs that are in this film
   */
  characters: z.array(z.string().url()),
  /**
   * The ISO 8601 date format of the time that this resource was created */
  created: z.string().datetime(),
  /**
   * The name of the director of this film */
  director: z.string(),
  /**
   * The ISO 8601 date format of the time that this resource was edited */
  edited: z.string().datetime(),
  /**
   * The episode number of this film */
  episode_id: z.number(),
  /**
   * The opening paragraphs at the beginning of this film */
  opening_crawl: z.string(),
  /**
   * An array of planet resource URLs that are in this film */
  planets: z.array(z.string().url()),
  /**
   * The name(s) of the producer(s) of this film. Comma separated */
  producer: z.string(),
  /**
   * The ISO 8601 date format of film release at original creator country */
  release_date: z.string(),
  /**
   * An array of species resource URLs that are in this film */
  species: z.array(z.string().url()),
  /**
   * An array of starship resource URLs that are in this film */
  starships: z.array(z.string().url()),
  /**
   * The title of this film */
  title: z.string(),
  /**
   * The hypermedia URL of this resource */
  url: z.string().url(),
  /**
   * An array of vehicle resource URLs that are in this film */
  vehicles: z.array(z.string().url()),
});

/**
 * @docs https://swapi.dev/documentation#films
 */
export const FilmsSchema = defineSWAPIPaginatedSchema(FilmSchema);

export const FilmParamsSchema = z.object({
  id: z.number(),
});
