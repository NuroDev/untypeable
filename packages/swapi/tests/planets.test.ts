import { describe, it, expect } from "vitest";

import { PlanetSchema, PlanetsSchema } from "../src/zod";
import { useTestClient } from "../src/_shared/_shared.util";

describe.concurrent("SWAPI - Planets", () => {
  const client = useTestClient();

  it("GET - /planets", async () => {
    const planets = await client("/planets");

    expect(planets).toBeDefined();
    expect(planets.count).toBeDefined();
    expect(planets.next).toBeDefined();
    expect(planets.previous).toBeDefined();
    expect(Array.isArray(planets.results)).toBe(true);
    expect(planets.results.length).toBeGreaterThan(0);
    expect(planets.results.at(0)).toBeDefined();

    expect(PlanetsSchema.safeParse(planets).success).toBe(true);
  });

  it("GET - /planets/:id", async () => {
    const planet = await client("/planets/:id", {
      id: 1,
    });

    expect(planet).toBeDefined();
    expect(planet).toMatchSnapshot({
      climate: "arid",
      created: "2014-12-09T13:50:49.641000Z",
      diameter: "10465",
      edited: "2014-12-20T20:58:18.411000Z",
      films: [
        "https://swapi.dev/api/films/1/",
        "https://swapi.dev/api/films/3/",
        "https://swapi.dev/api/films/4/",
        "https://swapi.dev/api/films/5/",
        "https://swapi.dev/api/films/6/",
      ],
      gravity: "1 standard",
      name: "Tatooine",
      orbital_period: "304",
      population: "200000",
      residents: [
        "https://swapi.dev/api/people/1/",
        "https://swapi.dev/api/people/2/",
        "https://swapi.dev/api/people/4/",
        "https://swapi.dev/api/people/6/",
        "https://swapi.dev/api/people/7/",
        "https://swapi.dev/api/people/8/",
        "https://swapi.dev/api/people/9/",
        "https://swapi.dev/api/people/11/",
        "https://swapi.dev/api/people/43/",
        "https://swapi.dev/api/people/62/",
      ],
      rotation_period: "23",
      surface_water: "1",
      terrain: "desert",
      url: "https://swapi.dev/api/planets/1/",
    });

    expect(PlanetSchema.safeParse(planet).success).toBe(true);
  });
});
