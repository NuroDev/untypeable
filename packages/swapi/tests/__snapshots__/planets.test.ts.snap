// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SWAPI - Planets > GET - /planets/:id 1`] = `
{
  "climate": "arid",
  "created": "2014-12-09T13:50:49.641000Z",
  "diameter": "10465",
  "edited": "2014-12-20T20:58:18.411000Z",
  "films": [
    "https://swapi.dev/api/films/1/",
    "https://swapi.dev/api/films/3/",
    "https://swapi.dev/api/films/4/",
    "https://swapi.dev/api/films/5/",
    "https://swapi.dev/api/films/6/",
  ],
  "gravity": "1 standard",
  "name": "Tatooine",
  "orbital_period": "304",
  "population": "200000",
  "residents": [
    "https://swapi.dev/api/people/1/",
    "https://swapi.dev/api/people/2/",
    "https://swapi.dev/api/people/4/",
    "https://swapi.dev/api/people/6/",
    "https://swapi.dev/api/people/7/",
    "https://swapi.dev/api/people/8/",
    "https://swapi.dev/api/people/9/",
    "https://swapi.dev/api/people/11/",
    "https://swapi.dev/api/people/43/",
    "https://swapi.dev/api/people/62/",
  ],
  "rotation_period": "23",
  "surface_water": "1",
  "terrain": "desert",
  "url": "https://swapi.dev/api/planets/1/",
}
`;
